City-of-Helsinki/drupal-helfi-tyollisyyspalvelut-manuaali
# TP-1128 Required fields issue

* Creates two new fields to replace two paragraph fields.
* Migrates content from the old fields to the new ones.
* Prevents sending mail during the update hook.
* Takes translations into accout, but due to the current content, there's no need to test this with translations.

**Testing instructions:**

**Before** pulling the branch, edit some existing services:
- [x] Make sure to have data at the "Ohjaus palveluun (Sisältö)" and "Merkinnät asiakastietojärjestelmään (Sisältö)" fields (located under the Specialist tab).
- [x] Make sure you have services at different states: some are _draft_, some are _ready to publish_, some are _published_.
- [x] And also make sure some services are first _published_ and then there is later version at _draft_ or _ready to publish_.

(It might be good idea to export the testing db at this point if you need to run the deploy multiple times.)

Pull the changes from this branch and run `composer install` and `drush deploy`.
- [x] Check that during the `deploy` command, no mail is sent.
- [x] Make sure the `deploy` command succeeds.
- [x] Check the services: make sure the new _"Ohjaus palveluun"_ and _"Merkinnät asiakastietojärjestelmään"_ fields have the old data at the **edit form**.
- [x] Check that also the serivces' **internal pages** shows the correct field data.
- [x] Make sure the old fields are hidden (at the edit form and at the internal page).

Also make sure the **revisions** are migrated correctly:
- [x] There shouldn't be any new revisions.
- [x] The revision states are not affected.
- [x] The _published_ revisions' data **does not override** data in _draft_ and _ready to publish_ revisions (e.g. the edit form and the internal page shows different field data).
- [x] The _draft_ and _ready to publish_ revisions' data **does not override** data at _published_ revisions (e.g. the edit form and the internal page shows different field data).

Also test the field requirement by editing the services:
- [x] Edit a service and leave the _"Ohjaus palveluun"_ and _"Merkinnät asiakastietojärjestelmään_" fields empty.
- [x] Make sure you can save the service as _ready to publish_ state (with a user who can access the Specialist tab).
- [x] Make sure the two fields are required when saving as published.

And also test the new prevent mail functionality:
- [x] Make sure you can access `/admin/config/system/prevent-sending-mail` with a user with an admin role.
- [x] Make sure you can't access the URL with anonymous or other users.
- [ ] Turn on "Block sending mail" and make sure (e.g. by requesting new password or by other test mail) that mails are **not send**.
- [x] Turn off the setting and make sure mails **are sent** normally.
- [x] Run the related tests: `lando test public/modules/custom/hel_tpm_general/tests/Kernel/PreventMailTest.php`

**Review checklist:**
